controllers/subscriptions.go

```diff
@@ -73,95 +73,134 @@ func (r *StorageSystemReconciler) ensureSubscriptions(instance *odfv1alpha1.Stor
 	}
 
 	for _, desiredSubscription := range subs {
-		err = r.CheckExistingSubscriptions(desiredSubscription)
-		if err != nil {
+		err = EnsureDesiredSubscription(r.Client, desiredSubscription)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			logger.Error(err, "failed to ensure subscription", "Subscription", desiredSubscription.Name)
 			return err
 		}
 
-		odfSub, err := GetOdfSubscription(r.Client)
+		err = instance.AddReferenceToRelatedObjects(r.Scheme, desiredSubscription)
 		if err != nil {
 			return err
 		}
+	}
 
-		// create/update subscription
-		sub := &operatorv1alpha1.Subscription{}
-		sub.ObjectMeta = desiredSubscription.ObjectMeta
-		_, err = controllerutil.CreateOrUpdate(context.TODO(), r.Client, sub, func() error {
-			sub.Spec = desiredSubscription.Spec
-			return controllerutil.SetControllerReference(odfSub, sub, r.Scheme)
-		})
-		if err != nil && !errors.IsAlreadyExists(err) {
-			logger.Error(err, "failed to create subscription")
-			return err
-		}
+	return nil
+}
 
-		err = instance.AddReferenceToRelatedObjects(r.Scheme, sub)
-		if err != nil {
-			return err
-		}
+func EnsureDesiredSubscription(cli client.Client, desiredSubscription *operatorv1alpha1.Subscription) error {
+
+	var err error
+
+	err = CheckExistingSubscriptions(cli, desiredSubscription)
+	if err != nil {
+		return err
+	}
+
+	// create/update subscription
+	sub := &operatorv1alpha1.Subscription{}
+	sub.ObjectMeta = desiredSubscription.ObjectMeta
+	_, err = controllerutil.CreateOrUpdate(context.TODO(), cli, sub, func() error {
+		sub.Spec = desiredSubscription.Spec
+		return SetOdfSubControllerReference(cli, sub)
+	})
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
 	}
 
 	return nil
 }
 
-func (r *StorageSystemReconciler) isVendorCsvReady(instance *odfv1alpha1.StorageSystem, logger logr.Logger) error {
+func GetVendorCsvNames(kind odfv1alpha1.StorageKind) []string {
 
 	var csvNames []string
 
-	if instance.Spec.Kind == VendorFlashSystemCluster() {
+	if kind == VendorFlashSystemCluster() {
 		csvNames = []string{IbmSubscriptionStartingCSV}
-	} else if instance.Spec.Kind == VendorStorageCluster() {
+	} else if kind == VendorStorageCluster() {
 		csvNames = []string{NoobaaSubscriptionStartingCSV, OcsSubscriptionStartingCSV}
 	}
 
-	for _, csvName := range csvNames {
+	return csvNames
+}
+
+func EnsureVendorCsv(cli client.Client, csvName string) (*operatorv1alpha1.ClusterServiceVersion, error) {
 
-		csvObj := &operatorv1alpha1.ClusterServiceVersion{}
-		err := r.Client.Get(context.TODO(), types.NamespacedName{
-			Name: csvName, Namespace: instance.Spec.Namespace}, csvObj)
+	var err error
 
+	csvObj := &operatorv1alpha1.ClusterServiceVersion{}
+	err = cli.Get(context.TODO(), types.NamespacedName{
+		Name: csvName, Namespace: OperatorNamespace}, csvObj)
+	if err != nil {
+		err = fmt.Errorf("CSV is not successfully installed")
```

Comment:
Let's say not found and add an existing error also instead of creating the new error